File: packages/python/src/ghagen/_yaml_config.py
"""Shared YAML loading utility."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from ruamel.yaml import YAML, YAMLError


def load_yaml_config(path: Path) -> dict[str, Any]:
    """Read and parse a YAML file, raising ``ValueError`` on parse errors."""
    yaml = YAML()
    try:
        with path.open() as f:
            data = yaml.load(f)
    except YAMLError as exc:
        raise ValueError(f"{path}: failed to parse YAML: {exc}") from exc
    if data is None:
        return {}
    if not isinstance(data, dict):
        raise ValueError(f"{path}: expected a YAML mapping at top level")
    return dict(data)

File: packages/typescript/src/_yaml-config.ts
/**
 * Shared YAML config loading utility.
 *
 * Wraps the `yaml` package with consistent error handling for use
 * across the config and pin subsystems.
 */

import { readFileSync } from "node:fs";
import { parse } from "yaml";

/** Read and parse a YAML config file. Throws on parse errors. */
export function loadYamlConfig(path: string): Record<string, unknown> {
  let text: string;
  try {
    text = readFileSync(path, "utf8");
  } catch (err) {
    throw new Error(`${path}: failed to read YAML file: ${(err as Error).message}`, { cause: err });
  }
  let data: unknown;
  try {
    data = parse(text);
  } catch (err) {
    throw new Error(`${path}: failed to parse YAML: ${(err as Error).message}`, { cause: err });
  }
  if (data === null || data === undefined) return {};
  if (typeof data !== "object" || Array.isArray(data)) {
    throw new Error(`${path}: expected a YAML mapping at top level`);
  }
  return data as Record<string, unknown>;
}
